pkg/detection/detect.go

```diff
@@ -83,13 +83,14 @@ func init() {
 				return false
 			}
 
-			data, err := io.ReadAll(r)
-			if err != nil {
+			var buf bytes.Buffer
+			if _, err := io.Copy(&buf, r); err != nil {
 				return false
 			}
+			data := buf.Bytes()
```

Comment:
I'm just curious how `buf.Bytes()` improves performance since `buf.Bytes()` anyway loads the entire content to memory. I thought the memory consumption is not so different from `io.ReadAll()`. Am I missing something?

Reply:
I wasn't sure so I benchmarked them in isolation to better understand. As far I can understand, there does seem to be a difference: https://github.com/simar7/copy-bench/blob/main/results.md (you can find the test code here: https://github.com/simar7/copy-bench/blob/main/main.go)

Having said that, I also tested passing in `buf` directly rather than `buf.Bytes()`, that further improves it:

### Change
```diff
-               if err := armjson.Unmarshal(data, &sniff, &metadata); err != nil {
+               if err := armjson.Unmarshal(buf, &sniff, &metadata); err != nil {
```

Where the buffer is passed in directly, rather than `[]byte`
```
func Unmarshal(dataBuf bytes.Buffer, target interface{}, metadata *types.Metadata) error {
       node, err := newParser(NewPeekReader(&dataBuf), Position{1, 1}).parse(metadata)
```

### Benchmark: 
#### new: is current PR
#### new2: is the change above.

```shell
benchstat new new2
goos: darwin
goarch: amd64
pkg: github.com/aquasecurity/defsec/pkg/detection
cpu: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
                   │     new      │                new2                │
                   │    sec/op    │   sec/op     vs base               │
IsType_SmallFile-8   3.778µ ± 20%   3.648µ ± 1%  -3.43% (p=0.041 n=10)
IsType_BigFile-8     2.014µ ± 29%   1.912µ ± 3%  -5.11% (p=0.000 n=10)
geomean              2.759µ         2.641µ       -4.27%

                   │     new      │                 new2                  │
                   │     B/op     │     B/op      vs base                 │
IsType_SmallFile-8   5.916Ki ± 0%   5.916Ki ± 0%       ~ (p=1.000 n=10) ¹
IsType_BigFile-8     9.644Ki ± 0%   9.644Ki ± 0%       ~ (p=1.000 n=10) ¹
geomean              7.553Ki        7.553Ki       +0.00%
¹ all samples are equal

                   │    new     │                new2                 │
                   │ allocs/op  │ allocs/op   vs base                 │
IsType_SmallFile-8   60.00 ± 0%   60.00 ± 0%       ~ (p=1.000 n=10) ¹
IsType_BigFile-8     10.00 ± 0%   10.00 ± 0%       ~ (p=1.000 n=10) ¹
geomean              24.49        24.49       +0.00%
¹ all samples are equal
```

Reply:
[The result](https://github.com/simar7/copy-bench/blob/main/results.md) shows almost the same memory allocation, doesn't it?

```
          │ big.readall │            big.copy             │
          │    B/op     │    B/op     vs base             │
BigFile-8    0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=100) ¹
¹ all samples are equal
```

I may be reading these results incorrectly. Please correct me if I'm wrong.

Reply:
Memory wise yes there's no difference but the issue is that with a large number of large files, trivy scan times out. This change improves the overall time it takes to run, as you can see in the results.

```
                   │  io.ReadAll  │               io.Copy               │
                   │    sec/op    │   sec/op     vs base                │
IsType_SmallFile-8   3.704µ ± 27%   3.707µ ± 1%        ~ (p=0.971 n=10)
IsType_BigFile-8     3.875µ ± 32%   1.973µ ± 2%  -49.07% (p=0.000 n=10)
geomean              3.788µ         2.705µ       -28.60%
```

This results shows that for small files the change is negligible (as expected). But for big files, with a sample set of n=10, it's considerably faster.